Stochastic Gradient Descent (SGD) is a popular optimization algorithm used in machine learning and deep learning. It’s widely used to minimize the cost function or loss function in models like neural networks.

 Gradient Descent Formula:
θ represents the parameters (weights) of the model
η is the learning rate, a hyperparameter that controls the step size.
∇f(θ)is the gradient of the function f with respect to 𝜃.


What is Stochastic Gradient Descent (SGD)?
In traditional Gradient Descent (also called Batch Gradient Descent), the algorithm computes the gradient of the cost function using the entire dataset. This can be computationally expensive and slow, especially for large datasets.

Stochastic Gradient Descent (SGD) is a variation of gradient descent where the gradient is computed using only a single data point (or a small batch) at each step. This makes the updates much faster but introduces more noise in the learning process, leading to more frequent updates that can sometimes help the model escape local minima.

θ=θ−η∇f(θ;x 
i
​
 ,y 
i
​
 )

f(θ;x 
i
​
 ,y 
i
​
 ) is the loss function for the single data point 
(
𝑥
𝑖
,
𝑦
𝑖
)
(x 
i
​
 ,y 
i
​
 ).

In [3]:
import tensorflow as tf 

#data \\\ training
x_train = tf.constant([[1.0],[2.0],[3.0],[4.0]], dtype=tf.float32)
y_train = tf.constant([[2.0],[4.0],[6.0],[8.0]],dtype=tf.float32)

#our model
model = tf.keras.Sequential([tf.keras.layers.Dense(1, input_shape=(1,))])
#compile model
model.compile(optimizer=tf.keras.optimizers.SGD(learning_rate = 0.01),loss="mean_squared_error")
#train model
model.fit(x_train,y_train , epochs=100 , verbose=2)


Epoch 1/100
1/1 - 0s - 170ms/step - loss: 97.6981
Epoch 2/100
1/1 - 0s - 27ms/step - loss: 67.8514
Epoch 3/100
1/1 - 0s - 15ms/step - loss: 47.1409
Epoch 4/100
1/1 - 0s - 17ms/step - loss: 32.7701
Epoch 5/100
1/1 - 0s - 17ms/step - loss: 22.7981
Epoch 6/100
1/1 - 0s - 17ms/step - loss: 15.8783
Epoch 7/100
1/1 - 0s - 16ms/step - loss: 11.0765
Epoch 8/100
1/1 - 0s - 15ms/step - loss: 7.7443
Epoch 9/100
1/1 - 0s - 17ms/step - loss: 5.4318
Epoch 10/100
1/1 - 0s - 17ms/step - loss: 3.8269
Epoch 11/100
1/1 - 0s - 16ms/step - loss: 2.7129
Epoch 12/100
1/1 - 0s - 15ms/step - loss: 1.9396
Epoch 13/100
1/1 - 0s - 16ms/step - loss: 1.4026
Epoch 14/100
1/1 - 0s - 17ms/step - loss: 1.0297
Epoch 15/100
1/1 - 0s - 16ms/step - loss: 0.7706
Epoch 16/100
1/1 - 0s - 17ms/step - loss: 0.5905
Epoch 17/100
1/1 - 0s - 22ms/step - loss: 0.4652
Epoch 18/100
1/1 - 0s - 18ms/step - loss: 0.3779
Epoch 19/100
1/1 - 0s - 17ms/step - loss: 0.3170
Epoch 20/100
1/1 - 0s - 18ms/step - loss: 0.2745
Epoch 21/100
1/1 - 0s

Epoch 1/100: Indicates the first epoch out of 100. The initial loss is very high at 97.6981, meaning the model's predictions are initially far from accurate.
Epoch Progress: As training continues, the loss decreases significantly. For example:
Epoch 2: Loss drops to 67.8514.
Epoch 10: Loss reduces further to 3.8269.
Epoch 50: Loss reaches 0.1507.
Epoch 100: Final loss is 0.1117

In [4]:
model.fit(x_train,y_train , epochs=10 , verbose=2)

Epoch 1/10
1/1 - 0s - 14ms/step - loss: 0.1110
Epoch 2/10
1/1 - 0s - 20ms/step - loss: 0.1103
Epoch 3/10
1/1 - 0s - 17ms/step - loss: 0.1097
Epoch 4/10
1/1 - 0s - 18ms/step - loss: 0.1090
Epoch 5/10
1/1 - 0s - 25ms/step - loss: 0.1084
Epoch 6/10
1/1 - 0s - 19ms/step - loss: 0.1077
Epoch 7/10
1/1 - 0s - 16ms/step - loss: 0.1071
Epoch 8/10
1/1 - 0s - 17ms/step - loss: 0.1064
Epoch 9/10
1/1 - 0s - 16ms/step - loss: 0.1058
Epoch 10/10
1/1 - 0s - 18ms/step - loss: 0.1052
